File: src/cpp/examples/backprop_last_layer.cc
/*
A demo for on-device backprop (transfer learning) of a classification model.

This demo runs a similar task as described in TF Poets tutorial, except that
learning happens on-device.
https://codelabs.developers.google.com/codelabs/tensorflow-for-poets/#0

Here are the steps to prepare the experiment data set:
  1) mkdir -p /tmp/retrain/
  2) curl http://download.tensorflow.org/example_images/flower_photos.tgz \
       | tar xz -C /tmp/retrain
  3) mogrify -format bmp /tmp/retrain/flower_photos//*//*.jpg

For more information, see
https://coral.ai/docs/edgetpu/retrain-classification-ondevice-backprop/
*/

#include <dirent.h>
#include <stdio.h>
#include <stdlib.h>

#include <random>

#include "absl/flags/flag.h"
#include "absl/flags/parse.h"
#include "glog/logging.h"
#include "src/cpp/basic/basic_engine.h"
#include "src/cpp/classification/engine.h"
#include "src/cpp/learn/backprop/softmax_regression_model.h"
#include "src/cpp/test_utils.h"

ABSL_FLAG(std::string, embedding_extractor_path,
          "/tmp/mobilenet_v1_1.0_224_quant_embedding_extractor_edgetpu.tflite",
          "Path to the embedding extractor tflite model.");
ABSL_FLAG(std::string, data_dir, "/tmp/retrain/flower_photos",
          "Path to the data set.");
ABSL_FLAG(std::string, output_model_path, "/tmp/trained_model_edgetpu.tflite",
          "Path to the output tflite model.");

namespace coral {
namespace learn {
namespace backprop {
namespace {

// Pair of class label and image path.
using LabelAndPath = std::pair<int, std::string>;

constexpr double kValidationDataRatio = 0.1;
constexpr double kTestDataRatio = 0.1;
constexpr int kNumTrainingIterations = 500;
constexpr int kBatchSize = 100;
constexpr int kPrintEvery = 100;

// Returns list of bmp file paths under the given folder.
std::vector<std::string> ListBmpFilesUnderDir(const std::string& parent_dir) {
  std::vector<std::string> file_list;
  file_list.reserve(4096);  // reserve enough elements for the flower dataset.
  DIR* dir = opendir(parent_dir.c_str());
  if (dir) {
    struct dirent* file = nullptr;
    while ((file = readdir(dir)) != nullptr) {
      if (strstr(file->d_name, ".bmp"))
        file_list.push_back(parent_dir + "/" + file->d_name);
    }
    closedir(dir);
  }
  LOG(INFO) << file_list.size() << " bmp files found in folder " << parent_dir;
  return file_list;
}

// Returns number of classes.
int ListFilesInSubdirs(const std::string& grandparent_dir,
                       std::vector<LabelAndPath>* label_and_paths) {
  CHECK(label_and_paths);
  DIR* dir = opendir(grandparent_dir.c_str());
  int label = 0;
  if (dir) {
    struct dirent* subdir = nullptr;
    while ((subdir = readdir(dir)) != nullptr) {
      if (subdir->d_type != DT_DIR) continue;
      if (!strcmp(subdir->d_name, ".")) continue;
      if (!strcmp(subdir->d_name, "..")) continue;
      // Read samples of a new class.
      LOG(INFO) << "Read samples from subfolder " << subdir->d_name
                << " for class label " << label;
      const auto files =
          ListBmpFilesUnderDir(grandparent_dir + "/" + subdir->d_name);
      for (const auto& p : files) {
        label_and_paths->push_back(std::make_pair(label, p));
      }
      ++label;
    }
    closedir(dir);
  }
  return label;
}

// Returns number of classes.
int SplitDataset(const std::string& data_dir,
                 std::vector<LabelAndPath>* training_label_and_paths,
                 std::vector<LabelAndPath>* validation_label_and_paths,
                 std::vector<LabelAndPath>* test_label_and_paths) {
  CHECK(training_label_and_paths);
  CHECK(validation_label_and_paths);
  CHECK(test_label_and_paths);
  std::random_device rd;
  std::mt19937 gen(rd());
  std::uniform_real_distribution<> dis(0, 1);
  std::vector<LabelAndPath> all_label_and_paths;
  const int num_classes = ListFilesInSubdirs(data_dir, &all_label_and_paths);
  for (const auto& p : all_label_and_paths) {
    const double r = dis(gen);
    if (r < kValidationDataRatio) {
      validation_label_and_paths->push_back(p);
    } else if (r < kValidationDataRatio + kTestDataRatio) {
      test_label_and_paths->push_back(p);
    } else {
      training_label_and_paths->push_back(p);
    }
  }
  LOG(INFO) << "Number of training samples: "
            << training_label_and_paths->size();
  LOG(INFO) << "Number of validation samples: "
            << validation_label_and_paths->size();
  LOG(INFO) << "Number of test samples: " << test_label_and_paths->size();
  return num_classes;
}

void ExtractEmbedding(BasicEngine* embedding_extractor,
                      const std::vector<LabelAndPath>& label_and_paths,
                      int num_classes, int feature_dim, Tensor* embeddings,
                      std::vector<int>* labels) {
  CHECK(embedding_extractor);
  CHECK(embeddings);
  CHECK(labels);

  embeddings->resize(label_and_paths.size(), feature_dim);
  labels->resize(label_and_paths.size(), -1);

  const auto& input_shape = embedding_extractor->get_input_tensor_shape();
  for (int r = 0; r < label_and_paths.size(); ++r) {
    const int label = label_and_paths[r].first;
    CHECK_LT(label, num_classes);
    const auto& image_path = label_and_paths[r].second;

    // Extract embedding vector.
    const std::vector<uint8_t> input_tensor = GetInputFromImage(
        image_path, {input_shape[1], input_shape[2], input_shape[3]});
    std::vector<std::vector<float>> results =
        embedding_extractor->RunInference(input_tensor);
    CHECK_EQ(results.size(), 1);
    const auto& embedding = results[0];
    CHECK_EQ(embedding.size(), feature_dim);
    embeddings->row(r) =
        Eigen::Map<const Eigen::VectorXf>(embedding.data(), embedding.size());

    // Create one-hot label vector.
    (*labels)[r] = label;
  }
}

// Returns classification accuracy.
float EvaluateTrainedModel(const std::string& model_path,
                           const std::vector<LabelAndPath>& label_and_paths) {
  // Load the model.
  coral::ClassificationEngine engine(model_path);
  std::vector<int> input_tensor_shape = engine.get_input_tensor_shape();
  int num_hits = 0;
  for (int r = 0; r < label_and_paths.size(); ++r) {
    // Read the image.
    const std::vector<uint8_t> input_tensor = coral::GetInputFromImage(
        label_and_paths[r].second,
        {input_tensor_shape[1], input_tensor_shape[2], input_tensor_shape[3]});
    // Classify without thresholding.
    auto results = engine.ClassifyWithInputTensor(
        input_tensor, /*threshold=*/-std::numeric_limits<float>::infinity(),
        /*top_k=*/1);
    CHECK_EQ(results.size(), 1);
    if (results[0].id == label_and_paths[r].first) ++num_hits;
  }
  return static_cast<float>(num_hits) / label_and_paths.size();
}

void TrainAndEvaluate(const std::string& embedding_extractor_path,
                      const std::string& data_dir,
                      const std::string& output_model_path) {
  BasicEngine embedding_extractor(embedding_extractor_path);
  const auto& output_sizes = embedding_extractor.get_all_output_tensors_sizes();
  CHECK_EQ(output_sizes.size(), 1);
  const int feature_dim = output_sizes[0];

  const auto& t0 = std::chrono::steady_clock::now();
  std::vector<LabelAndPath> training_label_and_paths,
      validation_label_and_paths, test_label_and_paths;
  const int num_classes =
      SplitDataset(data_dir, &training_label_and_paths,
                   &validation_label_and_paths, &test_label_and_paths);

  TrainingData training_data;
  ExtractEmbedding(&embedding_extractor, training_label_and_paths, num_classes,
                   feature_dim, &training_data.training_data,
                   &training_data.training_labels);
  ExtractEmbedding(&embedding_extractor, validation_label_and_paths,
                   num_classes, feature_dim, &training_data.validation_data,
                   &training_data.validation_labels);

  // Run training
  const auto& t1 = std::chrono::steady_clock::now();
  SoftmaxRegressionModel model(feature_dim, num_classes);
  model.Initialize();
  TrainConfig train_config = {kNumTrainingIterations, kBatchSize, kPrintEvery};
  SgdUpdater sgd_updater;
  model.Train(training_data, train_config, &sgd_updater);

  const auto& t2 = std::chrono::steady_clock::now();
  LOG(INFO)
      << "Time to get embedding vectors (ms): "
      << std::chrono::duration_cast<std::chrono::milliseconds>(t1 - t0).count();
  LOG(INFO)
      << "Time to train last layer (ms): "
      << std::chrono::duration_cast<std::chrono::milliseconds>(t2 - t1).count();

  // Append learned weights to input model and save as tflite format.
  model.SaveAsTfliteModel(embedding_extractor_path, output_model_path);

  // Evaluate the trained model.
  const float training_accuracy =
      EvaluateTrainedModel(output_model_path, training_label_and_paths);
  LOG(INFO) << "Accuracy on training data: " << training_accuracy;
  const float validation_accuracy =
      EvaluateTrainedModel(output_model_path, validation_label_and_paths);
  LOG(INFO) << "Accuracy on validation data: " << validation_accuracy;
  const float test_accuracy =
      EvaluateTrainedModel(output_model_path, test_label_and_paths);
  LOG(INFO) << "Accuracy on test data: " << test_accuracy;
}

}  // namespace
}  // namespace backprop
}  // namespace learn
}  // namespace coral

int main(int argc, char* argv[]) {
  absl::ParseCommandLine(argc, argv);
  coral::learn::backprop::TrainAndEvaluate(
      absl::GetFlag(FLAGS_embedding_extractor_path),
      absl::GetFlag(FLAGS_data_dir), absl::GetFlag(FLAGS_output_model_path));
}

File: examples/backprop_last_layer.py
# Copyright 2019 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
r"""A demo for on-device backprop (transfer learning) of a classification model.

This demo runs a similar task as described in TF Poets tutorial, except that
learning happens on-device.
https://codelabs.developers.google.com/codelabs/tensorflow-for-poets/#0

Here are the steps:
  1) mkdir -p /tmp/retrain/

  2) curl http://download.tensorflow.org/example_images/flower_photos.tgz \
       | tar xz -C /tmp/retrain

  3) Start training:

      python3 backprop_last_layer.py \
      --data_dir /tmp/retrain/flower_photos \
      --embedding_extractor_path \
      models/mobilenet_v1_1.0_224_quant_embedding_extractor_edgetpu.tflite

     Weights for retrained last layer will be saved to /tmp/retrain/output by
     default.

  4) Run an inference with the new model:

      python3 classify_image.py \
      --model /tmp/retrain/output/retrained_model_edgetpu.tflite \
      --label /tmp/retrain/output/label_map.txt
      --image images/sunflower.bmp

For more information, see
https://coral.ai/docs/edgetpu/retrain-classification-ondevice-backprop/
"""

import argparse
import contextlib
import os
import sys
import time

from edgetpu.basic import basic_engine
from edgetpu.learn.backprop.softmax_regression import SoftmaxRegression
import numpy as np
from PIL import Image


@contextlib.contextmanager
def test_image(path):
  """Returns opened test image."""
  with open(path, 'rb') as f:
    with Image.open(f) as image:
      yield image


def save_label_map(label_map, out_path):
  """Saves label map to a file."""
  with open(out_path, 'w') as f:
    for key, val in label_map.items():
      f.write('%s %s\n' % (key, val))


def get_image_paths(data_dir):
  """Walks through data_dir and returns list of image paths and label map.

  Args:
    data_dir: string, path to data directory. It assumes data directory is
      organized as,
          - [CLASS_NAME_0]
            -- image_class_0_a.jpg
            -- image_class_0_b.jpg
            -- ...
          - [CLASS_NAME_1]
            -- image_class_1_a.jpg
            -- ...
  Returns:
    A tuple of (image_paths, labels, label_map)
    image_paths: list of string, represents image paths
    labels: list of int, represents labels
    label_map: a dictionary (int -> string), e.g., 0->class0, 1->class1, etc.
  """
  classes = None
  image_paths = []
  labels = []

  class_idx = 0
  for root, dirs, files in os.walk(data_dir):
    if root == data_dir:
      # Each sub-directory in `data_dir`
      classes = dirs
    else:
      # Read each sub-directory
      assert classes[class_idx] in root
      print('Reading dir: %s, which has %d images' % (root, len(files)))
      for img_name in files:
        image_paths.append(os.path.join(root, img_name))
        labels.append(class_idx)
      class_idx += 1

  return image_paths, labels, dict(zip(range(class_idx), classes))


def shuffle_and_split(image_paths, labels, val_percent=0.1, test_percent=0.1):
  """Shuffles and splits data into train, validation, and test sets.

  Args:
    image_paths: list of string, of dim num_data
    labels: list of int of length num_data
    val_percent: validation data set percentage.
    test_percent: test data set percentage.

  Returns:
    Two dictionaries (train_and_val_dataset, test_dataset).
    train_and_val_dataset has the following fields.
      'data_train': data_train
      'labels_train': labels_train
      'data_val': data_val
      'labels_val': labels_val
    test_dataset has the following fields.
      'data_test': data_test
      'labels_test': labels_test
  """
  image_paths = np.array(image_paths)
  labels = np.array(labels)
  perm = np.random.permutation(image_paths.shape[0])
  image_paths = image_paths[perm]
  labels = labels[perm]

  num_total = image_paths.shape[0]
  num_val = int(num_total * val_percent)
  num_test = int(num_total * test_percent)
  num_train = num_total - num_val - num_test

  train_and_val_dataset = {}
  train_and_val_dataset['data_train'] = image_paths[0:num_train]
  train_and_val_dataset['labels_train'] = labels[0:num_train]
  train_and_val_dataset['data_val'] = image_paths[num_train:num_train + num_val]
  train_and_val_dataset['labels_val'] = labels[num_train:num_train + num_val]
  test_dataset = {}
  test_dataset['data_test'] = image_paths[num_train + num_val:]
  test_dataset['labels_test'] = labels[num_train + num_val:]
  return train_and_val_dataset, test_dataset


def extract_embeddings(image_paths, engine):
  """Uses model to process images as embeddings.

  Reads image, resizes and feeds to model to get feature embeddings. Original
  image is discarded to keep maximum memory consumption low.

  Args:
    image_paths: ndarray, represents a list of image paths.
    engine: BasicEngine, wraps embedding extractor model.

  Returns:
    ndarray of length image_paths.shape[0] of embeddings.
  """
  _, input_height, input_width, _ = engine.get_input_tensor_shape()
  assert engine.get_num_of_output_tensors() == 1
  feature_dim = engine.get_output_tensor_size(0)

  embeddings = np.empty((len(image_paths), feature_dim), dtype=np.float32)
  for idx, path in enumerate(image_paths):
    with test_image(path) as img:
      img = img.resize((input_width, input_height), Image.NEAREST)
      _, embeddings[idx, :] = engine.run_inference(np.asarray(img).flatten())

  return embeddings


def train(model_path, data_dir, output_dir):
  """Trains a softmax regression model given data and embedding extractor.

  Args:
    model_path: string, path to embedding extractor.
    data_dir: string, directory that contains training data.
    output_dir: string, directory to save retrained tflite model and label map.
  """
  t0 = time.perf_counter()
  image_paths, labels, label_map = get_image_paths(data_dir)
  train_and_val_dataset, test_dataset = shuffle_and_split(image_paths, labels)
  # Initializes basic engine model here to avoid repeatedly initialization,
  # which is time consuming.
  engine = basic_engine.BasicEngine(model_path)
  print('Extract embeddings for data_train')
  train_and_val_dataset['data_train'] = extract_embeddings(
      train_and_val_dataset['data_train'], engine)
  print('Extract embeddings for data_val')
  train_and_val_dataset['data_val'] = extract_embeddings(
      train_and_val_dataset['data_val'], engine)
  t1 = time.perf_counter()
  print('Data preprocessing takes %.2f seconds' % (t1 - t0))

  # Construct model and start training
  weight_scale = 5e-2
  reg = 0.0
  feature_dim = train_and_val_dataset['data_train'].shape[1]
  num_classes = np.max(train_and_val_dataset['labels_train']) + 1
  model = SoftmaxRegression(
      feature_dim, num_classes, weight_scale=weight_scale, reg=reg)

  learning_rate = 1e-2
  batch_size = 100
  num_iter = 500
  model.train_with_sgd(
      train_and_val_dataset, num_iter, learning_rate, batch_size=batch_size)
  t2 = time.perf_counter()
  print('Training takes %.2f seconds' % (t2 - t1))

  # Append learned weights to input model and save as tflite format.
  out_model_path = os.path.join(output_dir, 'retrained_model_edgetpu.tflite')
  model.save_as_tflite_model(model_path, out_model_path)
  print('Model %s saved.' % out_model_path)
  label_map_path = os.path.join(output_dir, 'label_map.txt')
  save_label_map(label_map, label_map_path)
  print('Label map %s saved.' % label_map_path)
  t3 = time.perf_counter()
  print('Saving retrained model and label map takes %.2f seconds' % (t3 - t2))

  retrained_engine = basic_engine.BasicEngine(out_model_path)
  test_embeddings = extract_embeddings(test_dataset['data_test'],
                                       retrained_engine)
  saved_model_acc = np.mean(
      np.argmax(test_embeddings, axis=1) == test_dataset['labels_test'])
  print('Saved tflite model test accuracy: %.2f%%' % (saved_model_acc * 100))
  t4 = time.perf_counter()
  print('Checking test accuracy takes %.2f seconds' % (t4 - t3))


def main():
  parser = argparse.ArgumentParser()
  parser.add_argument(
      '--embedding_extractor_path',
      required=True,
      help='Path to embedding extractor tflite model.')
  parser.add_argument('--data_dir', required=True, help='Directory to data.')
  parser.add_argument(
      '--output_dir',
      default='/tmp/retrain/output',
      help='Path to directory to save retrained model and label map.')
  args = parser.parse_args()

  if not os.path.exists(args.data_dir):
    sys.exit('%s does not exist!' % args.data_dir)

  if not os.path.exists(args.output_dir):
    os.makedirs(args.output_dir)

  train(args.embedding_extractor_path, args.data_dir, args.output_dir)


if __name__ == '__main__':
  main()
